### ISSUE TITLE
FileManipulator error when using remote file

### ISSUE THREAD

**Issue by kyle-mccarthy** [spatie/laravel-medialibrary#121]
Hi,

I am in the process of debugging an issue that arises when attempting to execute conversions on a remote image.  I am using Rackspace for the file storage.

Currently, the original image (non-manipulated) image will upload, but then when manipulations are attempted an error will occur.  So far I have found that this is a result of the temp file inside storage/medialibrary/temp being empty (size 0).

When the FileManipulator calls executes the following code in performConversions

```
app(Filesystem::class)->copyFromMediaLibrary($media, $copiedOriginalFile);
```

The value of $copiedOriginalFile is the broken empty file.  Therefore, when the conversions are iterated over and the $copiedOriginalFile is passed the error occurs since it is size 0.

```
$stream = $this->filesystem->disk($media->disk)->readStream($sourceFile);
```

The $stream when using the rackspace file system will return FALSE when attempting to read (stream_get_contents) the stream indicating that the EOF has been reached.  

I have attempted reading the stream when using the local filesystem and stream_get_contents returns the contents of the file as expected. 

The error is as follows:

```
[2015-12-15 00:30:19] local.ERROR: exception 'Intervention\Image\Exception\NotReadableException' with message 'Unable to init from given binary data.' in /home/vagrant/Code/vendor/intervention/image/src/Intervention/Image/Gd/Decoder.php:92
Stack trace:
#0 /home/vagrant/Code/vendor/intervention/image/src/Intervention/Image/AbstractDecoder.php(299): Intervention\Image\Gd\Decoder->initFromBinary('')
#1 /home/vagrant/Code/vendor/intervention/image/src/Intervention/Image/AbstractDriver.php(64): Intervention\Image\AbstractDecoder->init('')
#2 /home/vagrant/Code/vendor/intervention/image/src/Intervention/Image/ImageManager.php(50): Intervention\Image\AbstractDriver->init('')
#3 /home/vagrant/Code/vendor/league/glide/src/Api/Api.php(85): Intervention\Image\ImageManager->make('')
#4 /home/vagrant/Code/vendor/spatie/laravel-glide/src/Spatie/Glide/GlideImage.php(102): League\Glide\Api\Api->run(Object(Symfony\Component\HttpFoundation\Request), '')
#5 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/FileManipulator.php(110): Spatie\Glide\GlideImage->save('/home/vagrant/C...')
#6 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/FileManipulator.php(77): Spatie\MediaLibrary\FileManipulator->performConversion(Object(Spatie\MediaLibrary\Media), Object(Spatie\MediaLibrary\Conversion\Conversion), '/home/vagrant/C...')
#7 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/FileManipulator.php(49): Spatie\MediaLibrary\FileManipulator->performConversions(Object(Spatie\MediaLibrary\Conversion\ConversionCollection), Object(Spatie\MediaLibrary\Media))
#8 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/Filesystem.php(54): Spatie\MediaLibrary\FileManipulator->createDerivedFiles(Object(Spatie\MediaLibrary\Media))
#9 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/FileAdder/FileAdder.php(318): Spatie\MediaLibrary\Filesystem->add('/tmp/php4pZx5l', Object(Spatie\MediaLibrary\Media), 'logo.png')
#10 /home/vagrant/Code/vendor/spatie/laravel-medialibrary/src/FileAdder/FileAdder.php(254): Spatie\MediaLibrary\FileAdder\FileAdder->toCollectionOnDisk('posts', '')
#11 /home/vagrant/Code/app/Http/Controllers/PostController.php(85): Spatie\MediaLibrary\FileAdder\FileAdder->toMediaLibrary('posts')
#12 [internal function]: app\Http\Controllers\PostController->postCreate(Object(Illuminate\Http\Request))
#13 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(256): call_user_func_array(Array, Array)
#14 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(164): Illuminate\Routing\Controller->callAction('postCreate', Array)
#15 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(112): Illuminate\Routing\ControllerDispatcher->call(Object(app\Http\Controllers\PostController), Object(Illuminate\Routing\Route), 'postCreate')
#16 [internal function]: Illuminate\Routing\ControllerDispatcher->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#17 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(139): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#18 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#19 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#20 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(114): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#21 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(69): Illuminate\Routing\ControllerDispatcher->callWithinStack(Object(app\Http\Controllers\PostController), Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'postCreate')
#22 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Route.php(203): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request), 'App\\Http\\Contro...', 'postCreate')
#23 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Route.php(134): Illuminate\Routing\Route->runWithCustomDispatcher(Object(Illuminate\Http\Request))
#24 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Router.php(708): Illuminate\Routing\Route->run(Object(Illuminate\Http\Request))
#25 [internal function]: Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#26 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(139): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#27 /home/vagrant/Code/app/Http/Middleware/Authenticate.php(45): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#28 [internal function]: app\Http\Middleware\Authenticate->handle(Object(Illuminate\Http\Request), Object(Closure))
#29 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#30 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#31 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#32 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Router.php(710): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#33 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Router.php(675): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#34 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Routing/Router.php(635): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#35 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(236): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#36 [internal function]: Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#37 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(139): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#38 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(50): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#39 [internal function]: Illuminate\Foundation\Http\Middleware\VerifyCsrfToken->handle(Object(Illuminate\Http\Request), Object(Closure))
#40 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#41 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#42 [internal function]: Illuminate\View\Middleware\ShareErrorsFromSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#43 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#44 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(62): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#45 [internal function]: Illuminate\Session\Middleware\StartSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#46 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#47 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#48 [internal function]: Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(Object(Illuminate\Http\Request), Object(Closure))
#49 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#50 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(59): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#51 [internal function]: Illuminate\Cookie\Middleware\EncryptCookies->handle(Object(Illuminate\Http\Request), Object(Closure))
#52 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#53 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(44): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#54 [internal function]: Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#55 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(124): call_user_func_array(Array, Array)
#56 [internal function]: Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#57 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): call_user_func(Object(Closure), Object(Illuminate\Http\Request))
#58 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(122): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#59 /home/vagrant/Code/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(87): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#60 /home/vagrant/Code/public/index.php(54): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#61 {main}  
```

**Comment by kyle-mccarthy** [comment]
I am not entirely sure if this is an issue with the medialibrary, if it is from packages used in openshift, or if I am just overlooking something.  But here are some things that I have found out:

1. $this->filesystem->disk($media->disk)->has($sourceFile) returns true
2. $this->filesystem->disk($media->disk)->read($sourceFile) will return the array with the correct contents
3. $this->filesystem->disk($media->disk)->readStream($sourceFile) returns the following array:

```
array:7 [▼
  "type" => "file"
  "dirname" => "1"
  "path" => "1/logo.png"
  "timestamp" => 1450153645
  "mimetype" => "image/png"
  "size" => "14497"
  "stream" => stream resource @57 ▼
    wrapper_type: "PHP"
    stream_type: "TEMP"
    mode: "w+b"
    unread_bytes: 0
    seekable: true
    uri: "php://temp"
    options: []
  }
]
```

However, file_put_contents will result in an empty/broken file when using readStream.  I have tried reading the data directly from the readStream implemented in the RackspaceAdapter class from league/flysystem-rackspace, but I haven't had any luck getting the correct results.

**Comment by freekmurze** [comment]
I did not come across this error before. I've only used the local and S3 drivers myself. Are you using [the latest adapter](https://github.com/thephpleague/flysystem-rackspace)?

**Comment by kyle-mccarthy** [comment]
I was using the latest adapter.  I ended up ripping out Rackspace and using S3 instead, in general it has been much easier to work with.  I spent all of yesterday attempting to narrow down the issue but was only able to get it to work by using the filesystem read function instead of readStream. 

Specifically in the copyFromMediaLibrary method of the Filesystem class contained in the your package, I changed it to $stream = $this->filesystem->disk($media->disk)->read($sourceFile); rather than $stream = $this->filesystem->disk($media->disk)->readStream($sourceFile);.  However, this was more of a hack than an actual fix, especially since using the read function will gobble up memory.
